ur_robot_driver/src/ros/hardware_interface.cpp

```diff
@@ -431,15 +435,18 @@ void HardwareInterface::read(const ros::Time& time, const ros::Duration& period)
   }
   else
   {
-    ROS_ERROR("Could not get fresh data package from robot");
+    if (!non_blocking_read_)
+    {
+      ROS_ERROR("Could not get fresh data package from robot");
+    }
   }
 }
 
 void HardwareInterface::write(const ros::Time& time, const ros::Duration& period)
 {
   if ((runtime_state_ == static_cast<uint32_t>(rtde_interface::RUNTIME_STATE::PLAYING) ||
        runtime_state_ == static_cast<uint32_t>(rtde_interface::RUNTIME_STATE::PAUSING)) &&
-      robot_program_running_)
+      robot_program_running_ && ((non_blocking_read_ && packet_read_) || (!non_blocking_read_)))
```

Comment:
Could this be reduced to 
```suggestion
      robot_program_running_ && packet_read_
```

If we are using blocking read then we still set `packet_read_` to `true` in each cycle, right?

Reply:
If I've understood correctly, that would force read and write to be synchronised whether or not the non_blocking_read parameter was set. At the moment there's nothing that forces one write for every one read. When used with combined_robot_hw we get eight writes (probably containing the same data) for every one read due to the arm running at 125Hz and the other robot in the loop running at 1KHz. It could be simplified to: 
```suggestion
      robot_program_running_ && (!non_blocking_read_ || packet_read_))
```
but I thought the current logic was a little more readable.

Reply:
Although looking at them next to each other, I do prefer ` robot_program_running_ && (!non_blocking_read_ || packet_read_))
`

Reply:
I also like your second suggestion the most.

Reply:
https://github.com/UniversalRobots/Universal_Robots_ROS_Driver/pull/86/commits/3a8eea838eec04d5fae6405b4135f251a2a1a6ba